mono/utils/mono-os-mutex.h

```diff
@@ -246,51 +246,64 @@ mono_os_cond_broadcast (mono_cond_t *cond)
 
 #else
 
-typedef CRITICAL_SECTION mono_mutex_t;
+// FIXME mono_mutex_t and mono_mutex_recursive_t.
+typedef struct mono_mutex_t {
+	union {
+		CRITICAL_SECTION critical_section;
+		SRWLOCK srwlock;
+	};
+	gboolean recursive;
+} mono_mutex_t;
+
 typedef CONDITION_VARIABLE mono_cond_t;
 
 static inline void
 mono_os_mutex_init (mono_mutex_t *mutex)
 {
-	BOOL res;
-
-	res = InitializeCriticalSectionEx (mutex, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
-	if (G_UNLIKELY (res == 0))
-		g_error ("%s: InitializeCriticalSectionEx failed with error %d", __func__, GetLastError ());
+	mutex->recursive = FALSE;
+	InitializeSRWLock (&mutex->srwlock);
 }
 
 static inline void
 mono_os_mutex_init_recursive (mono_mutex_t *mutex)
 {
-	BOOL res;
+	mutex->recursive = TRUE;
+	const BOOL res = InitializeCriticalSectionEx (&mutex->critical_section, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
 
-	res = InitializeCriticalSectionEx (mutex, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
 	if (G_UNLIKELY (res == 0))
 		g_error ("%s: InitializeCriticalSectionEx failed with error %d", __func__, GetLastError ());
 }
 
 static inline void
 mono_os_mutex_destroy (mono_mutex_t *mutex)
 {
-	DeleteCriticalSection (mutex);
+	// There is no way to destroy a Win32 SRWLOCK.
+	if (mutex->recursive)
+		DeleteCriticalSection (&mutex->critical_section);
 }
 
 static inline void
 mono_os_mutex_lock (mono_mutex_t *mutex)
 {
-	EnterCriticalSection (mutex);
+	mutex->recursive ?
+		EnterCriticalSection (&mutex->critical_section) :
+		AcquireSRWLockExclusive (&mutex->srwlock);
 }
 
 static inline int
 mono_os_mutex_trylock (mono_mutex_t *mutex)
 {
-	return TryEnterCriticalSection (mutex) == 0 ? -1 : 0;
+	return (mutex->recursive ?
+		TryEnterCriticalSection (&mutex->critical_section) :
+		TryAcquireSRWLockExclusive (&mutex->srwlock)) ? 0 : -1;
```

Comment:
Btw, I forgot to check and mention -- this requires Windows 7.
i.e. not Vista.

Reply:
That's fine, we only support Win7+.